client/cherami/outputhostconnection.go

```diff
@@ -175,7 +173,23 @@ func (conn *outputHostConnection) isClosed() bool {
 	return atomic.LoadInt32(&conn.closed) != 0
 }
 
+// drainReadPipe reads and discards all messages on
+// the outputHostStream until it encounters
+// a read stream error
+func (conn *outputHostConnection) drainReadPipe() {
+	for {
+		if _, err := conn.outputHostStream.Read(); err != nil {
+			return
+		}
```

Comment:
@aravindvs - The trouble is - we close the ack tchannel immediately after calling consumer.close(). So, even if you give the opportunity for the app to process them, they won't be able to ack/nack them, so, all that would do is delay the shutdown. @samar & I discussed a proposal for supporting graceful shutdown on the client. It would involve adding another API, something like client.initiateClose(). When the app calls this, the library will drain the read channel and enqueue everything to the delivery channel (and subsequently also close the deliveryChannel). The closing of delivery channel will be the signal for the app to indicate EOF. The app then acks/nacks all of them and finally calls client.close(). 

This patch doesn't address the clean shutdown.